packages/expo-modules-core/android/src/main/cpp/JavaScriptWeakObject.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+#include "JavaScriptWeakObject.h"
+#include "JSIInteropModuleRegistry.h"
+
+namespace expo {
+
+void JavaScriptWeakObject::registerNatives() {
+  registerHybrid({
+      makeNativeMethod("lock", JavaScriptWeakObject::lock),
+  });
+}
+
+jni::local_ref<JavaScriptObject::javaobject> JavaScriptWeakObject::lock() {
+  jsi::Runtime &rt = _runtimeHolder.getJSRuntime();
+
+  std::shared_ptr<jsi::Object> objectPtr;
+  if (_weakObjectType == WeakObjectType::JSIWeakObject) {
+    jsi::Value value =
+        std::static_pointer_cast<jsi::WeakObject>(_weakObject)->lock(rt);
+    if (value.isUndefined()) {
+      return nullptr;
+    }
+    objectPtr = std::make_shared<jsi::Object>(value.asObject(rt));
+  } else if (_weakObjectType == WeakObjectType::WeakRef) {
+    objectPtr =
+        derefWeakRef(rt, std::static_pointer_cast<jsi::Object>(_weakObject));
+  } else {
+    objectPtr = std::static_pointer_cast<jsi::Object>(_weakObject);
+  }
+
+  if (!objectPtr) {
+    return nullptr;
+  }
+  return JavaScriptObject::newInstance(_runtimeHolder.getModuleRegistry(),
+                                       _runtimeHolder, objectPtr);
+}
+
+jni::local_ref<jni::HybridClass<JavaScriptWeakObject, Destructible>::javaobject>
+JavaScriptWeakObject::newInstance(
+    JSIInteropModuleRegistry *jsiInteropModuleRegistry,
+    std::weak_ptr<JavaScriptRuntime> runtime,
+    std::shared_ptr<jsi::Object> jsObject) {
+  auto weakObject = JavaScriptWeakObject::newObjectCxxArgs(std::move(runtime),
+                                                           std::move(jsObject));
+  jsiInteropModuleRegistry->jniDeallocator->addReference(weakObject);
+  return weakObject;
+}
+
+JavaScriptWeakObject::JavaScriptWeakObject(
+    WeakRuntimeHolder runtime, std::shared_ptr<jsi::Object> jsObject)
+    : _runtimeHolder(std::move(runtime)) {
+  _runtimeHolder.ensureRuntimeIsValid();
+  jsi::Runtime &rt = _runtimeHolder.getJSRuntime();
+
+  try {
+    _weakObject = std::make_shared<jsi::WeakObject>(rt, *jsObject);
+    _weakObjectType = WeakObjectType::JSIWeakObject;
+    return;
+  } catch (const std::logic_error &) {
+    // JSCRuntime will throw std::logic_error from unimplemented jsi::WeakObject
+  }
+
+  // Check whether the runtime supports `WeakRef` objects. If it does not,
+  // we consciously hold a strong reference to the object and cause memory
+  // leaks.
+  if (isWeakRefSupported(rt)) {
+    _weakObject = createWeakRef(rt, jsObject);
+    _weakObjectType = WeakObjectType::WeakRef;
+  } else {
+    _weakObject = jsObject;
+    _weakObjectType = WeakObjectType::NotSupported;
```

Comment:
Reality speaking, can that happen? We have some form of weak objects on all supported runtimes (JSC, Hermes, V8). Throwing an error rather than masking an unsupported operation is better. Otherwise, we may not notice something missing on new runtime in the future, causing massive leaks.

Reply:
that's a good question. in fact, on android, [jsc supports jsi::WeakObject](https://github.com/facebook/react-native/blob/f6b7243d2ae7c8d7fe690da8f9a922df844061d3/packages/react-native/ReactCommon/jsc/CMakeLists.txt#L33-L35). it's just ios < 14.5 would go into this NonSupported flow.
on react-native 0.74, meta removed the flag and jsc should support jsi::WeakObject in all cases. maybe we can remove the logic for WeakRefAvailable/NonSupported here and use jsi::WeakObject for everything on android.

when we moved to 0.74, ios could remove the WeakRef code and align with android implementations. does that make sense to you?

ps. i didn't add jsi::WeakObject to v8 yet which i should. maybe it's good to update it in the christmas seasons.

Reply:
i somehow changed my mind. since android supports jsi::WeakObject for all cases. i would go a step further to remove the WeakRef fallback code. ultimately on 0.74, ios will do the same removal and we don't need the shared WeakRef fallback code at all.